packages/core/test/actions.test.ts

```diff
@@ -847,6 +847,254 @@ describe('entry/exit actions', () => {
 
       expect(actual).toEqual(['child_b', 'b', 'child_a', 'a', 'root']);
     });
+
+    it('an exit action executed when an interpreter gets stopped should receive `xstate.stop` event', () => {
+      let receivedEvent;
+      const machine = createMachine({
+        exit: (_ctx, ev) => {
+          receivedEvent = ev;
+        }
+      });
+
+      const service = interpret(machine).start();
+      service.stop();
+
+      expect(receivedEvent).toEqual({ type: 'xstate.stop' });
+    });
+
+    // https://github.com/statelyai/xstate/issues/2880
+    it('stopping an interpreter that receives events from its children exit handlers should not throw', () => {
+      const child = createMachine({
+        id: 'child',
+        initial: 'idle',
+        states: {
+          idle: {
+            exit: sendParent('EXIT')
+          }
+        }
+      });
+
+      const parent = createMachine({
+        id: 'parent',
+        invoke: child
+      });
+
+      const interpreter = interpret(parent);
+      interpreter.start();
+
+      expect(() => interpreter.stop()).not.toThrow();
+    });
+
+    it('send actions from exit handlers of a stopped child should not be executed', () => {
+      const child = createMachine({
+        id: 'child',
+        initial: 'idle',
+        states: {
+          idle: {
+            exit: sendParent('EXIT')
+          }
+        }
+      });
+
+      const parent = createMachine({
+        id: 'parent',
+        context: () => ({
+          child: spawn(child)
+        }),
+        on: {
+          STOP_CHILD: {
+            actions: stop((ctx: any) => ctx.child)
+          },
+          EXIT: {
+            actions: () => {
+              throw new Error('This should not be called.');
+            }
+          }
+        }
+      });
+
+      const interpreter = interpret(parent).start();
+      interpreter.send({ type: 'STOP_CHILD' });
+    });
+
+    it('actors spawned in exit handlers of a stopped child should not be started', () => {
```

Comment:
Likewise, this test should still pass by virtue of _not_ spawning any actor after being stopped, rather than spawning a deferred actor

Reply:
This would be problematic because some actions might refer to the "spawned" actor, like for example here:
```js
exit: [
  assign({
    childRef: () => spawn(child)
  }),
  send({ type: 'GRAB_SOME_DATA_FROM_ME' }, { to: ctx => ctx.childRef })
]
```

So we need to create **something** within this `spawn` call but also we shouldn't really `spawn` a real child. This also brings up the fact that we should introduce a warning like this to more places:
https://github.com/statelyai/xstate/blob/11d815ee92044e0fb8f4ce89a6e6b2f6cf8dd2b0/packages/core/src/interpreter.ts#L1416-L1422

Ideally, we'd always receive "lazy" entities from users within `assign`. This doesn't hold true right now and we allow eager entities but we should probably require all of them to be lazy. To avoid breaking changes we still need to process those eager entities but that should be accompanied by a dev-only warning. Thoughts?

Reply:
Yes, let's add a dev-only warning.

In v5 (the spawn improvements branch), this is fixed by making all actors "not started" at first, and then starting them with their given behavior, while using the exact same reference. All actors are lazy.